InterruptServer.rb

```diff
@@ -0,0 +1,233 @@
+require 'socket'
+require 'json'
+
+class InterruptServer
+
+  CHAT_LENGTH = 45 # length of chat text user can see at one time
+  TICK_LENGTH = 2 # how many seconds to wait before 'moving chat text left'
+  MAX_MSG_LENGTH = 1024 # max length of incoming message read
+  
+  def initialize(host, port, num_buckets=5)
+    @server = UDPSocket.new
+    @server.bind(host, port)
+
+    puts "running on host: #{host}"
+
+    @clients = {}
+    @buckets_available = [*0..num_buckets-1]
+    # buckets: at most one client can be assigned a given bucket, though clients don't have to be assigned a bucket
+    # the intended use on the client side is to use this to assign colors to users
+
+    @chat_array = Array.new(CHAT_LENGTH, [' ', nil]) # each element is [character, bucket]
+    @outbox = [] # array of messages, as hashes with sender info
+  end
+
+  def run
+    loop do
+      incoming = IO.select([@server], nil, nil, TICK_LENGTH)
+      if incoming.nil?
+        tick
+      else
+        data, sender = @server.recvfrom(MAX_MSG_LENGTH)
+        handle_incoming(data, sender)
+      end
+      handle_outbox
+    end
+  end
+
+  def tick
+    if @chat_array != ' ' * CHAT_LENGTH # TODO: update this
+      update_chat_array(' ', nil)
+    end
+  end
+
+  def handle_incoming(data, sender)
+    msg = parse_msg(data)
+    return if msg == nil
+
+    key, host, port = sender_info(sender)
+
+    case msg['type']
+    when 'connect'
+      add_client(key, host, port, msg['name'], msg['time'])
+      assign_bucket(key)
+      ack_client(host, port)
+
+    when 'quit'
+      delete_client(key)
+
+    when 'chat'
+      if @clients.has_key? key and new_msg?(key, msg['time'])
+        update_time(key, msg['time'])
+        update_chat_array(msg['body'], key)
+
+      end
+    end
+  end
+
+  def assign_bucket(client_key)
+    return if @buckets_available.empty?
+
+    @clients[client_key]['bucket'] = @buckets_available.shift
+  end
+
+  def handle_outbox
+    while not @outbox.empty?
+      msg = @outbox.shift
+
+      case msg['type']
+      when 'chat'
+        @clients.each{ |key, client|
+          host = client['host']
+          port = client['port']
+          send_msg(msg['msg'], host, port)
+        }
+
+      when 'private_new'
+        send_msg(msg['msg'], msg['host'], msg['port'])
+
+      when 'private'
+        if @clients.has_key?(msg['key'])
+          client = @clients[key]
+          send_msg(msg['msg'], client['host'], client['port'])
+        end
+
+      end
+    end
+  end
+
+  def sender_info(sender)
+    port = sender[1]
+    host = sender[2]
+    key = port.to_s + host # host + port is unique per client
+
+    [key, host, port]
+  end
+
+  # parse to json, and ensure formatted correctly
+  # return nil if not
+  def parse_msg(data)
+    begin
+      msg = JSON.parse(data)
+      msg = check_fields(msg)
+    rescue
+      nil
+    end
+  end
+
+  # return nil if fields wrong, otherwise return msg
+  def check_fields(msg)
```

Comment:
this is also ripe to be an object. It feels like it wants to be something like `Message#valid?`.